tsl/src/compression/compression.c

```diff
@@ -2267,93 +2746,93 @@ decompress_batches_for_insert(const ChunkInsertState *cis, TupleTableSlot *slot)
+																  slot,
+																  &index_rel,
+																  &num_index_scankeys);
+
+	if (index_rel)
 	{
-		bool valid = true;
-
-		/*
-		 * Since the heap scan API does not support SK_SEARCHNULL we have to check
-		 * for NULL values manually when those are part of the constraints.
-		 */
-		for (int attno = bms_next_member(null_columns, -1); attno >= 0;
-			 attno = bms_next_member(null_columns, attno))
+		/* Prepare the heap scan keys if any */
+		Bitmapset *filtered_key_columns = NULL;
+		int i = -1;
+		while ((i = bms_next_member(key_columns, i)) > 0)
 		{
-			if (!slot_attisnull(compressed_slot, attno))
+			AttrNumber attno = i + FirstLowInvalidHeapAttributeNumber;
+			char *attname = get_attname(out_rel->rd_id, attno, false);
+			if (!ts_array_is_member(settings->fd.segmentby, attname))
 			{
-				valid = false;
-				break;
+				filtered_key_columns = bms_add_member(filtered_key_columns, i);
 			}
 		}
+		int num_scankeys;
+		ScanKeyData *scankeys = build_scankeys(cis->hypertable_relid,
+											   in_rel,
+											   out_rel,
+											   settings,
+											   filtered_key_columns,
+											   &null_columns,
+											   slot,
+											   &num_scankeys);
+		bms_free(key_columns);
 
 		/*
-		 * Skip if NULL check failed.
+		 * Using latest snapshot to scan the heap since we are doing this to build
+		 * the index on the uncompressed chunks in order to do speculative insertion
+		 * which is always built from all tuples (even in higher levels of isolation).
 		 */
-		if (!valid)
-			continue;
-
-		bool should_free;
-		HeapTuple tuple = ExecFetchSlotHeapTuple(compressed_slot, false, &should_free);
-		heap_deform_tuple(tuple,
-						  decompressor.in_desc,
-						  decompressor.compressed_datums,
-						  decompressor.compressed_is_nulls);
-
-		if (should_free)
-			heap_freetuple(tuple);
-
-		write_logical_replication_msg_decompression_start();
-		row_decompressor_decompress_row_to_table(&decompressor);
-
-		TM_FailureData tmfd;
-		TM_Result result pg_attribute_unused();
-		result = table_tuple_delete(in_rel,
-									&compressed_slot->tts_tid,
-									decompressor.mycid,
-									snapshot,
-									InvalidSnapshot,
-									true,
-									&tmfd,
-									false);
-
-		write_logical_replication_msg_decompression_end();
-
-		Assert(result == TM_Ok);
-
-		Assert(cis->cds != NULL);
-		cis->cds->batches_decompressed += decompressor.batches_decompressed;
-		cis->cds->tuples_decompressed += decompressor.tuples_decompressed;
+		stats = decompress_batches_indexscan(in_rel,
+											 out_rel,
+											 index_rel,
+											 GetLatestSnapshot(),
+											 index_scankeys,
+											 num_index_scankeys,
+											 scankeys,
+											 num_scankeys,
+											 NULL, /* no null column check for non-segmentby
+													  columns */
+											 NIL);
+		index_close(index_rel, AccessShareLock);
+	}
+	else
+	{
+		int num_scankeys;
+		ScanKeyData *scankeys = build_scankeys(cis->hypertable_relid,
```

Comment:
Let's add `heap` to these guys as well, i.e. `num_heap_scankeys`, `heap_scankeys`, `build_heap_scankeys`.